zero-downtime-schema-migrations-postgresql.mdx

```diff
@@ -0,0 +1,119 @@
+---
+title: 'How to perform Postgres schema changes in production with zero downtime'
+description: "Schema migrations are a pain. In this workshop we'll review the problem space and tools available to make reversible schema changes in production with zero downtime."
+image:
+  src: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/images/pgroll-internal.jpeg
+  alt: Zero-downtime schema migrations in PostgreSQL
+author: Alex Francoeur
+date: 01-31-2024
+tags: ['workshop', 'pgroll']
+published: true
+slug: zero-downtime-schema-migrations-postgresql
+---
+
+<ArticleVideo platform="youtube" src="https://www.youtube.com/embed/-1aO6UznfI0?si=qLv09ZanA_UMQf0w" />
+
+[Schema migrations are a pain](https://xata.io/blog/postgres-schema-changes-pita). They typically require multi-step playbooks to roll out in production and migrate your data over. In this workshop we review the problem space and tools available to make reversible schema changes in production with zero downtime.
+
+## Common pitfalls to rolling out schema changes
+
+There are a number of gotchas we typically see folks run into when trying to roll out schema changes in a production environment where downtime is simply not acceptable.
+
+**Additive-only changes and schema debt**
+
+If you’ve ever worked with a production database, I’m sure you’ve run into this one. We’ve interviewed many developers and architects about schema migrations and the one thing they’ll all admit is that they’d rather add columns than remove them. It’s the safe play, destructive changes can be complex, but things can get out of hand very quickly. A single column can very quickly turn into multiple versions with different variations.
+
+This pattern comes from fear of breaking the application, which is fair. Though at the end up the day, having a table with hundred of columns can lead to a few issues. You can end up with performance implications returning certain queries, having long living compatibility code in your application and general confusion for anyone new to the database.
+
+**The locking minefield**
+
+If you aren’t familiar with Postgres, locking is a way to control concurrent access to data in tables. Locks are used pretty heavily when modifying your schema and [there are many types of different locks](https://www.postgresql.org/docs/current/explicit-locking.html) available in postgres.
+
+As an example, any ALTER TABLE statements will require the ACCESS EXCLUSIVE lock which conflicts with all other lock types.
```

Comment:
This line should be combined with the one below; it's got a hard newline at the end right now.